avl.3
.Dd 2001-06-10
.Dt DEBFOSTER 8
.Os "Debian GNU/Linux"
.Sh NAME
.Nm avl
.Nd functions to create, manipulate and destroy augmented AVL trees
.Sh SYNOPSYS
.Li #include <avl.h>

/* User supplied function to compare two items like strcmp() does.
 * For example: cmp(a,b) will return:
 *   -1  if a < b
 *    0  if a = b
 *    1  if a > b
 */
typedef int (*avl_compare_t)(const void *, const void *);

/* User supplied function to delete an item when a node is free()d.
 * If NULL, the item is not free()d.
 */
typedef void (*avl_freeitem_t)(void *);

#define AVL_CMP(a,b) ((a) < (b) ? -1 : (a) > (b) ? 1 : 0)

#if defined(AVL_COUNT) && defined(AVL_DEPTH)
#define AVL_NODE_INIT(item) { 0, 0, 0, 0, 0, (item), 0, 0 }
#else
#define AVL_NODE_INIT(item) { 0, 0, 0, 0, 0, (item), 0 }
#endif

typedef struct avl_node_t {
	struct avl_node_t *next;
	struct avl_node_t *prev;
	struct avl_node_t *parent;
	struct avl_node_t *left;
	struct avl_node_t *right;
	void *item;
#ifdef AVL_COUNT
	unsigned long count;
#endif
#ifdef AVL_DEPTH
	unsigned char depth;
#endif
} avl_node_t;

extern const avl_node_t avl_node_0;

#define AVL_TREE_INIT(cmp,free) { 0, 0, 0, (cmp), (free) }

typedef struct avl_tree_t {
	avl_node_t *head;
	avl_node_t *tail;
	avl_node_t *top;
	avl_compare_t cmp;
	avl_freeitem_t freeitem;
} avl_tree_t;

extern const avl_tree_t avl_tree_0;

/* Initializes a new tree for elements that will be ordered using
 * the supplied strcmp()-like function.
 * Returns the value of avltree (even if it's NULL).
 * O(1) */
extern avl_tree_t *avl_tree_init(avl_tree_t *avltree, avl_compare_t, avl_freeitem_t);

/* Allocates and initializes a new tree for elements that will be
 * ordered using the supplied strcmp()-like function.
 * Returns NULL if memory could not be allocated.
 * O(1) */
extern avl_tree_t *avl_tree_alloc(avl_compare_t, avl_freeitem_t);

/* Frees the entire tree efficiently. Nodes will be free()d.
 * If the tree's freeitem is not NULL it will be invoked on every item.
 * O(n) */
extern void avl_tree_free(avl_tree_t *);

/* Reinitializes the tree structure for reuse. Nothing is free()d.
 * Compare and freeitem functions are left alone.
 * O(1) */
extern void avl_tree_clear(avl_tree_t *);

/* Free()s all nodes in the tree but leaves the tree itself.
 * If the tree's freeitem is not NULL it will be invoked on every item.
 * O(n) */
extern void avl_tree_purge(avl_tree_t *);

/* Initializes memory for use as a node. Returns NULL if avlnode is NULL.
 * O(1) */
extern avl_node_t *avl_node_init(avl_node_t *avlnode, void *item);

/* Insert an item into the tree and return the new node.
 * Returns NULL and sets errno if memory for the new node could not be
 * allocated or if the node is already in the tree (EEXIST).
 * O(lg n) */
extern avl_node_t *avl_item_insert(avl_tree_t *, void *item);

/* Insert an item into the tree and return the new node.
 * If a nodes with equal items are already in the tree, this item will
 * be inserted to the left of all those. Returns NULL and sets errno if
 * memory for the new node could not be allocated.
 * O(lg n) */
extern avl_node_t *avl_item_insert_left(avl_tree_t *, void *);

/* Insert an item into the tree and return the new node.
 * If a nodes with equal items are already in the tree, this item will
 * be inserted to the right of all those. Returns NULL and sets errno if
 * memory for the new node could not be allocated.
 * O(lg n) */
extern avl_node_t *avl_item_insert_right(avl_tree_t *, void *);

/* Insert an item into the tree and return the new node.
 * If a nodes with equal items are already in the tree, this item will
 * be inserted somewhere among those. Returns NULL and sets errno if
 * memory for the new node could not be allocated.
 * O(lg n) */
extern avl_node_t *avl_item_insert_somewhere(avl_tree_t *, void *item);

/* Insert an item before another node. Returns the new node.
 * If old is NULL, the item is appended to the tree.
 * Returns NULL and sets errno if memory for the new node could not be
 * allocated.
 * O(lg n) */
extern avl_node_t *avl_item_insert_before(avl_tree_t *, avl_node_t *old, void *new);

/* Insert an item after another node. Returns the new node.
 * If old is NULL, the item is prepended to the tree.
 * Returns NULL and sets errno if memory for the new node could not be
 * allocated.
 * O(lg n) */
extern avl_node_t *avl_item_insert_after(avl_tree_t *, avl_node_t *old, void *new);

/* Insert a node into the tree and return it.
 * Returns NULL if an equal node is already in the tree.
 * O(lg n) */
extern avl_node_t *avl_insert(avl_tree_t *, avl_node_t *);

/* Insert a node into the tree and return it.
 * If a nodes with equal items are already in the tree, this node will
 * be inserted to the left of all those.
 * O(lg n) */
extern avl_node_t *avl_insert_left(avl_tree_t *, avl_node_t *);

/* Insert a node into the tree and return it.
 * If a nodes with equal items are already in the tree, this node will
 * be inserted to the right of all those.
 * O(lg n) */
extern avl_node_t *avl_insert_right(avl_tree_t *, avl_node_t *);

/* Insert a node into the tree and return it.
 * If a nodes with equal items are already in the tree, this node will
 * be inserted somewhere among those.
 * O(lg n) */
extern avl_node_t *avl_insert_somewhere(avl_tree_t *, avl_node_t *);

/* Insert a node before another node. Returns the new node.
 * If old is NULL, the item is appended to the tree.
 * O(lg n) */
extern avl_node_t *avl_insert_before(avl_tree_t *, avl_node_t *old, avl_node_t *new);

/* Insert a node after another node. Returns the new node.
 * If old is NULL, the item is prepended to the tree.
 * O(lg n) */
extern avl_node_t *avl_insert_after(avl_tree_t *, avl_node_t *old, avl_node_t *new);

/* Deletes a node from the tree. Returns immediately if the node is NULL.
 * The item will not be free()d regardless of the tree's freeitem handler.
 * This function comes in handy if you need to update the search key.
 * O(lg n) */
extern void avl_unlink(avl_tree_t *, avl_node_t *);

/* Deletes a node from the tree. Returns immediately if the node is NULL.
 * If the tree's freeitem is not NULL, it is invoked on the item.
 * If it is, returns the item.
 * O(lg n) */
extern void *avl_delete(avl_tree_t *, avl_node_t *);

/* Searches for an item in the tree and deletes it if found.
 * If the tree's freeitem is not NULL, it is invoked on the item.
 * If it is, returns the item.
 * O(lg n) */
extern void *avl_item_delete(avl_tree_t *, const void *item);

/* If exactly one node is moved in memory, this will fix the pointers
 * in the tree that refer to it. It must be an exact shallow copy.
 * Returns the pointer to the old position.
 * O(1) */
extern avl_node_t *avl_fixup(avl_tree_t *, avl_node_t *new);

/* Searches for an item, returning either the first (leftmost) exact
 * match, or (if no exact match could be found) the first (leftmost)
 * of the nodes that have an item greater than the search item.
 * If exact is not NULL, *exact will be set to:
 *    0  if the returned node is inequal or NULL
 *    1  if the returned node is equal
 * Returns NULL if no equal or greater element could be found.
 * O(lg n) */
extern avl_node_t *avl_search_left(const avl_tree_t *, const void *item, int *exact);

/* Searches for an item, returning either the last (rightmost) exact
 * match, or (if no exact match could be found) the last (rightmost)
 * of the nodes that have an item smaller than the search item.
 * If exact is not NULL, *exact will be set to:
 *    0  if the returned node is inequal or NULL
 *    1  if the returned node is equal
 * Returns NULL if no equal or smaller element could be found.
 * O(lg n) */
extern avl_node_t *avl_search_right(const avl_tree_t *, const void *item, int *exact);

/* Searches for an item, returning either some exact
 * match, or (if no exact match could be found) the last (rightmost)
 * of the nodes that have an item smaller than the search item.
 * If exact is not NULL, *exact will be set to:
 *    0  if the returned node is inequal or NULL
 *    1  if the returned node is equal
 * Returns NULL if no equal or smaller element could be found.
 * O(lg n) */
extern avl_node_t *avl_search_rightish(const avl_tree_t *, const void *item, int *exact);

/* Searches for the item in the tree and returns a matching node if found
 * or NULL if not.
 * O(lg n) */

NEWS
Version 2.0.0 2009-09-29 Wessel Dankers <wsl@fruit.je>
	API change: names are now object-verb.
	Change build system to auto-fu+libtool
	Add manpages

Version 0.3.5 2002-11-15 Wessel Dankers <wsl@fruit.eu.org>
	Added avl_node_fixup() and avl_clear_tree().
	Removed obsolete files from source tree.

Version 0.3.4 2002-06-11 Wessel Dankers <wsl@fruit.eu.org>
	Fixed a bug in the node counting.

Version 0.3.1 2001-07-17 Wessel Dankers <wsl@nl.linux.org>
	Node initialization is moved to make matters a bit more robust

Version 0.3.0 2001-01-07 Wessel Dankers <wsl@nl.linux.org>
	Tree can now be balanced on count or depth (but depth works better).
	Fixed bug: balancing on count now is done correctly.

Version 0.2.0 2000-11-28 Wessel Dankers <wsl@nl.linux.org>
	Modifications to support fast traversal and accessing by index.
	The tree is generalized to allow arbitrary comparison functions.
	Fixed bug: when deleting, in some cases rebalancing would not occur

Version 0.1.0 (alpha):
	This is the initial alpha version of AVLTree.  It's already fully
	functional for many applications, including the one that I developed it
	for.  I've only tested it on my Linux 2.0.35/glibc2 system, so I have no
	idea what it will do anywhere else so far.  Let me know if you have good
	results or bad if you try a platform that I don't mention above.

	This version is considered alpha because it does not yet contain
	all of the features that I plan for version 1.0.0.  It should not
	contain any bugs as it is.

$Id$
$URL$


src/avl.c
/*****************************************************************************

	avl.c - Source code for libavl

	Copyright (c) 1998  Michael H. Buselli <cosine@cosine.org>
	Copyright (c) 2000-2009  Wessel Dankers <wsl@fruit.je>

	This file is part of libavl.

	libavl is free software: you can redistribute it and/or modify
	it under the terms of the GNU Lesser General Public License as
	published by the Free Software Foundation, either version 3 of
	the License, or (at your option) any later version.

	libavl is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Lesser General Public License for more details.

	You should have received a copy of the GNU General Public License
	and a copy of the GNU Lesser General Public License along with
	libavl.  If not, see <http://www.gnu.org/licenses/>.

	Augmented AVL-tree. Original by Michael H. Buselli <cosine@cosine.org>.

	Modified by Wessel Dankers <wsl@fruit.je> to add a bunch of bloat
	to the sourcecode, change the interface and replace a few bugs.
	Mail him if you find any new bugs.

	$Id$
	$URL$

*****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#include "avl.h"

static void avl_rebalance(avl_tree_t *, avl_node_t *);

#ifdef AVL_COUNT
#define NODE_COUNT(n)  ((n) ? (n)->count : 0)
#define L_COUNT(n)     (NODE_COUNT((n)->left))
#define R_COUNT(n)     (NODE_COUNT((n)->right))
#define CALC_COUNT(n)  (L_COUNT(n) + R_COUNT(n) + 1)
#endif

#ifdef AVL_DEPTH
#define NODE_DEPTH(n)  ((n) ? (n)->depth : 0)
#define L_DEPTH(n)     (NODE_DEPTH((n)->left))
#define R_DEPTH(n)     (NODE_DEPTH((n)->right))
#define CALC_DEPTH(n)  ((unsigned char)((L_DEPTH(n)>R_DEPTH(n)?L_DEPTH(n):R_DEPTH(n)) + 1))
#endif

const avl_node_t avl_node_0 = {0};
const avl_tree_t avl_tree_0 = {0};

#ifdef CAST_QUAL_KLUDGES
static avl_node_t *avl_const_node(const avl_node_t *node) {
	union {
		const avl_node_t *c;
		avl_node_t *v;
	} u;
	u.c = node;
	return u.v;
}

static void *avl_const_item(const void *item) {
	union {
		const void *c;
		void *v;
	} u;
	u.c = item;
	return u.v;
}
#else
#define avl_const_node(x) ((avl_node_t *)(x))
#define avl_const_item(x) ((void *)(x))
#endif

static int avl_check_balance(avl_node_t *avlnode) {
#ifdef AVL_DEPTH
	int d;
	d = R_DEPTH(avlnode) - L_DEPTH(avlnode);
	return d<-1?-1:d>1?1:0;
#else
/*	int d;
 *	d = ffs(R_COUNT(avlnode)) - ffs(L_COUNT(avlnode));
 *	d = d<-1?-1:d>1?1:0;
 */
#ifdef AVL_COUNT
	int pl, r;

	pl = ffs(L_COUNT(avlnode));
	r = R_COUNT(avlnode);

	if(r>>pl+1)
		return 1;
	if(pl<2 || r>>pl-2)
		return 0;
	return -1;
#else
#error No balancing possible.
#endif
#endif
}

#ifdef AVL_COUNT
unsigned long avl_count(const avl_tree_t *avltree) {
	if(!avltree)
		return 0;
	return NODE_COUNT(avltree->top);
}

avl_node_t *avl_at(const avl_tree_t *avltree, unsigned long index) {
	avl_node_t *avlnode;
	unsigned long c;

	if(!avltree)
		return NULL;

	avlnode = avltree->top;

	while(avlnode) {
		c = L_COUNT(avlnode);

		if(index < c) {
			avlnode = avlnode->left;
		} else if(index > c) {
			avlnode = avlnode->right;
			index -= c+1;
		} else {
			return avlnode;
		}
	}
	return NULL;
}

unsigned long avl_index(const avl_node_t *avlnode) {
	avl_node_t *next;
	unsigned long c;

	if(!avlnode)
		return 0;

	c = L_COUNT(avlnode);

	while((next = avlnode->parent)) {
		if(avlnode == next->right)
			c += L_COUNT(next) + 1;
		avlnode = next;
	}

	return c;
}
#endif

static const avl_node_t *avl_search_leftmost_equal(const avl_tree_t *tree, const avl_node_t *node, const void *item) {
	avl_compare_t cmp = tree->cmp;
	const avl_node_t *r = node;

	for(;;) {
		for(;;) {
			node = node->left;
			if(!node)
				return r;
			if(cmp(item, node->item))
				break;
			r = node;
		}
		for(;;) {
			node = node->right;
			if(!node)
				return r;
			if(!cmp(item, node->item))
				break;
		}
		r = node;
	}
}

static const avl_node_t *avl_search_rightmost_equal(const avl_tree_t *tree, const avl_node_t *node, const void *item) {
	avl_compare_t cmp = tree->cmp;
	const avl_node_t *r = node;

	for(;;) {
		for(;;) {
			node = node->right;
			if(!node)
				return r;
			if(cmp(item, node->item))
				break;
			r = node;
		}
		for(;;) {
			node = node->left;
			if(!node)
				return r;
			if(!cmp(item, node->item))
				break;
		}
		r = node;
	}
}

avl_node_t *avl_search_rightish(const avl_tree_t *tree, const void *item, int *exact) {
	avl_node_t *node;
	avl_compare_t cmp;
	int c;

	if(!exact)
		exact = &c;

	if(!tree)
		return *exact = 0, (avl_node_t *)NULL;

	node = tree->top;

	if(!node)
		return *exact = 0, (avl_node_t *)NULL;

	cmp = tree->cmp;

	for(;;) {
		c = cmp(item, node->item);

		if(c < 0) {
			if(node->left)
				node = node->left;
			else
				return *exact = 0, node->prev;
		} else if(c > 0) {
			if(node->right)
				node = node->right;
			else
				return *exact = 0, node;
		} else {
			return *exact = 1, node;
		}
	}
}

avl_node_t *avl_search_left(const avl_tree_t *tree, const void *item, int *exact) {
	avl_node_t *node;
	int c;

	if(!exact)
		exact = &c;

	if(!tree)
		return *exact = 0, (avl_node_t *)NULL;

	node = avl_search_rightish(tree, item, exact);

	if(*exact)
		return avl_const_node(avl_search_leftmost_equal(tree, node, item));

	if(node)
		return node->next;

	return tree->head;
}

avl_node_t *avl_search_right(const avl_tree_t *tree, const void *item, int *exact) {
	const avl_node_t *node;
	int c;

	if(!exact)
		exact = &c;

	node = avl_search_rightish(tree, item, exact);
	if(*exact)
		node = avl_search_rightmost_equal(tree, node, item);

	return avl_const_node(node);
}

avl_node_t *avl_search(const avl_tree_t *avltree, const void *item) {
	int c;
	avl_node_t *n;
	n = avl_search_rightish(avltree, item, &c);
	return c ? n : NULL;
}

avl_tree_t *avl_tree_init(avl_tree_t *avltree, avl_compare_t cmp, avl_freeitem_t freeitem) {
	if(avltree) {
		avltree->head = NULL;
		avltree->tail = NULL;
		avltree->top = NULL;
		avltree->cmp = cmp;
		avltree->freeitem = freeitem;
	}
	return avltree;
}

avl_tree_t *avl_tree_alloc(avl_compare_t cmp, avl_freeitem_t freeitem) {
	return avl_tree_init(malloc(sizeof(avl_tree_t)), cmp, freeitem);
}

avl_tree_t *avl_tree_clear(avl_tree_t *avltree) {
	if(avltree)
		avltree->top = avltree->head = avltree->tail = NULL;
	return avltree;
}

avl_tree_t *avl_tree_purge(avl_tree_t *avltree) {
	avl_node_t *node, *next;
	avl_freeitem_t freeitem;

	if(!avltree)
		return NULL;

	freeitem = avltree->freeitem;

	for(node = avltree->head; node; node = next) {
		next = node->next;
		if(freeitem)
			freeitem(node->item);
		free(node);
	}

	return avl_tree_clear(avltree);
}

void avl_tree_free(avl_tree_t *avltree) {
	if(!avltree)
		return;
	(void)avl_tree_purge(avltree);
	free(avltree);
}

static void avl_node_clear(avl_node_t *newnode) {
	newnode->left = newnode->right = NULL;
#	ifdef AVL_COUNT
	newnode->count = 1;
#	endif
#	ifdef AVL_DEPTH
	newnode->depth = 1;
#	endif
}

avl_node_t *avl_node_init(avl_node_t *newnode, const void *item) {
	if(newnode)
		newnode->item = avl_const_item(item);
	return newnode;
}

avl_node_t *avl_node_alloc(const void *item) {
	avl_node_t *newnode;
	newnode = malloc(sizeof *newnode);
	if(newnode)
		newnode->item = avl_const_item(item);
	return newnode;
}

/* Insert a node in an empty tree. If avlnode is NULL, the tree will be
 * cleared and ready for re-use.
 * If the tree is not empty, the old nodes are left dangling.
 * O(1) */
static avl_node_t *avl_insert_top(avl_tree_t *avltree, avl_node_t *newnode) {
	avl_node_clear(newnode);
	newnode->prev = newnode->next = newnode->parent = NULL;
	avltree->head = avltree->tail = avltree->top = newnode;
	return newnode;
}

avl_node_t *avl_insert_before(avl_tree_t *avltree, avl_node_t *node, avl_node_t *newnode) {
	if(!avltree || !newnode)
		return NULL;

	if(!node)
		return avltree->tail
			? avl_insert_after(avltree, avltree->tail, newnode)
			: avl_insert_top(avltree, newnode);

	if(node->left)
		return avl_insert_after(avltree, node->prev, newnode);

	avl_node_clear(newnode);

	newnode->next = node;
	newnode->parent = node;

	newnode->prev = node->prev;
	if(node->prev)
		node->prev->next = newnode;
	else
		avltree->head = newnode;
	node->prev = newnode;

	node->left = newnode;
	avl_rebalance(avltree, node);
	return newnode;
}

avl_node_t *avl_insert_after(avl_tree_t *avltree, avl_node_t *node, avl_node_t *newnode) {
	if(!avltree || !newnode)
		return NULL;

	if(!node)
		return avltree->head
			? avl_insert_before(avltree, avltree->head, newnode)
			: avl_insert_top(avltree, newnode);

	if(node->right)
		return avl_insert_before(avltree, node->next, newnode);

	avl_node_clear(newnode);

	newnode->prev = node;
	newnode->parent = node;

	newnode->next = node->next;
	if(node->next)
		node->next->prev = newnode;
	else
		avltree->tail = newnode;
	node->next = newnode;

	node->right = newnode;
	avl_rebalance(avltree, node);
	return newnode;
}

avl_node_t *avl_insert(avl_tree_t *avltree, avl_node_t *newnode) {
	avl_node_t *node;
	int c;

	node = avl_search_rightish(avltree, newnode->item, &c);
	return c ? NULL : avl_insert_after(avltree, node, newnode);
}

avl_node_t *avl_insert_left(avl_tree_t *avltree, avl_node_t *newnode) {
	return avl_insert_before(avltree, avl_search_left(avltree, newnode->item, NULL), newnode);
}

avl_node_t *avl_insert_right(avl_tree_t *avltree, avl_node_t *newnode) {
	return avl_insert_after(avltree, avl_search_right(avltree, newnode->item, NULL), newnode);
}

avl_node_t *avl_insert_somewhere(avl_tree_t *avltree, avl_node_t *newnode) {
	return avl_insert_after(avltree, avl_search_rightish(avltree, newnode->item, NULL), newnode);
}

avl_node_t *avl_item_insert(avl_tree_t *avltree, const void *item) {
	avl_node_t *newnode;

	if(!avltree)
		return errno = EFAULT, (avl_node_t *)NULL;

	newnode = avl_node_init(malloc(sizeof *newnode), item);
	if(newnode) {
		if(avl_insert(avltree, newnode))
			return newnode;
		free(newnode);
		errno = EEXIST;
	}
	return NULL;
}

avl_node_t *avl_item_insert_somewhere(avl_tree_t *avltree, const void *item) {
	avl_node_t *newnode;

	if(!avltree)
		return errno = EFAULT, (avl_node_t *)NULL;

	newnode = avl_node_init(malloc(sizeof *newnode), item);
	if(newnode)
		return avl_insert_somewhere(avltree, newnode);
	return NULL;
}

avl_node_t *avl_item_insert_before(avl_tree_t *avltree, avl_node_t *node, const void *item) {
	avl_node_t *newnode;

	if(!avltree)
		return errno = EFAULT, (avl_node_t *)NULL;

	newnode = avl_node_init(malloc(sizeof *newnode), item);
	if(newnode)
		return avl_insert_before(avltree, node, newnode);
	return NULL;
}

avl_node_t *avl_item_insert_after(avl_tree_t *avltree, avl_node_t *node, const void *item) {
	avl_node_t *newnode;

	if(!avltree)
		return errno = EFAULT, (avl_node_t *)NULL;

	newnode = avl_node_init(malloc(sizeof *newnode), item);
	if(newnode)
		return avl_insert_after(avltree, node, newnode);
	return NULL;
}

avl_node_t *avl_item_insert_left(avl_tree_t *avltree, const void *item) {
	avl_node_t *newnode;

	if(!avltree)
		return errno = EFAULT, (avl_node_t *)NULL;

	newnode = avl_node_init(malloc(sizeof *newnode), item);
	if(newnode)
		return avl_insert_left(avltree, newnode);
	return NULL;
}

avl_node_t *avl_item_insert_right(avl_tree_t *avltree, const void *item) {
	avl_node_t *newnode;

	if(!avltree)
		return errno = EFAULT, (avl_node_t *)NULL;

	newnode = avl_node_init(malloc(sizeof *newnode), item);
	if(newnode)
		return avl_insert_right(avltree, newnode);
	return NULL;
}

avl_node_t *avl_unlink(avl_tree_t *avltree, avl_node_t *avlnode) {
	avl_node_t *parent;
	avl_node_t **superparent;
	avl_node_t *subst, *left, *right;
	avl_node_t *balnode;

	if(!avltree || !avlnode)
		return NULL;

	if(avlnode->prev)
		avlnode->prev->next = avlnode->next;
	else
		avltree->head = avlnode->next;

	if(avlnode->next)
		avlnode->next->prev = avlnode->prev;
	else
		avltree->tail = avlnode->prev;

	parent = avlnode->parent;

	superparent = parent
		? avlnode == parent->left ? &parent->left : &parent->right
		: &avltree->top;

	left = avlnode->left;
	right = avlnode->right;
	if(!left) {
		*superparent = right;
		if(right)
			right->parent = parent;
		balnode = parent;
	} else if(!right) {
		*superparent = left;
		left->parent = parent;
		balnode = parent;
	} else {
		subst = avlnode->prev;
		if(subst == left) {
			balnode = subst;
		} else {
			balnode = subst->parent;
			balnode->right = subst->left;
			if(balnode->right)
				balnode->right->parent = balnode;
			subst->left = left;
			left->parent = subst;
		}
		subst->right = right;
		subst->parent = parent;
		right->parent = subst;
		*superparent = subst;
	}

	avl_rebalance(avltree, balnode);

	return avlnode;
}

void *avl_delete(avl_tree_t *avltree, avl_node_t *avlnode) {
	void *item = NULL;
	if(avlnode) {
		item = avlnode->item;
		(void)avl_unlink(avltree, avlnode);
		if(avltree->freeitem)
			avltree->freeitem(item);
		free(avlnode);
	}
	return item;
}

void *avl_item_delete(avl_tree_t *avltree, const void *item) {
	return avl_delete(avltree, avl_search(avltree, item));
}

avl_node_t *avl_fixup(avl_tree_t *avltree, avl_node_t *newnode) {
	avl_node_t *oldnode = NULL, *node;

	if(!avltree || !newnode)
		return NULL;

	node = newnode->prev;
	if(node) {
		oldnode = node->next;
		node->next = newnode;
	} else {
		avltree->head = newnode;
	}

	node = newnode->next;
	if(node) {
		oldnode = node->prev;
		node->prev = newnode;
	} else {
		avltree->tail = newnode;
	}

	node = newnode->parent;
	if(node) {
		if(node->left == oldnode)
			node->left = newnode;
		else
			node->right = newnode;
	} else {
		oldnode = avltree->top;
		avltree->top = newnode;
	}

	return oldnode;
}

/*
 * avl_rebalance:
 * Rebalances the tree if one side becomes too heavy.  This function
 * assumes that both subtrees are AVL-trees with consistant data.  The
 * function has the additional side effect of recalculating the count of
 * the tree at this node.  It should be noted that at the return of this
 * function, if a rebalance takes place, the top of this subtree is no
 * longer going to be the same node.
 */
static void avl_rebalance(avl_tree_t *avltree, avl_node_t *avlnode) {
	avl_node_t *child;
	avl_node_t *gchild;
	avl_node_t *parent;
	avl_node_t **superparent;

	parent = avlnode;

	while(avlnode) {
		parent = avlnode->parent;

		superparent = parent
			? avlnode == parent->left ? &parent->left : &parent->right
			: &avltree->top;

		switch(avl_check_balance(avlnode)) {
		case -1:
			child = avlnode->left;
#			ifdef AVL_DEPTH
			if(L_DEPTH(child) >= R_DEPTH(child)) {
#			else
#			ifdef AVL_COUNT
			if(L_COUNT(child) >= R_COUNT(child)) {
#			else
#			error No balancing possible.
#			endif
#			endif
				avlnode->left = child->right;
				if(avlnode->left)
					avlnode->left->parent = avlnode;
				child->right = avlnode;
				avlnode->parent = child;
				*superparent = child;
				child->parent = parent;
#				ifdef AVL_COUNT
				avlnode->count = CALC_COUNT(avlnode);
				child->count = CALC_COUNT(child);
#				endif
#				ifdef AVL_DEPTH
				avlnode->depth = CALC_DEPTH(avlnode);
				child->depth = CALC_DEPTH(child);
#				endif
			} else {
				gchild = child->right;
				avlnode->left = gchild->right;
				if(avlnode->left)
					avlnode->left->parent = avlnode;
				child->right = gchild->left;
				if(child->right)
					child->right->parent = child;
				gchild->right = avlnode;
				if(gchild->right)
					gchild->right->parent = gchild;
				gchild->left = child;
				if(gchild->left)
					gchild->left->parent = gchild;
				*superparent = gchild;
				gchild->parent = parent;
#				ifdef AVL_COUNT
				avlnode->count = CALC_COUNT(avlnode);
				child->count = CALC_COUNT(child);
				gchild->count = CALC_COUNT(gchild);
#				endif
#				ifdef AVL_DEPTH
				avlnode->depth = CALC_DEPTH(avlnode);
				child->depth = CALC_DEPTH(child);
				gchild->depth = CALC_DEPTH(gchild);
#				endif
			}
		break;
		case 1:
			child = avlnode->right;
#			ifdef AVL_DEPTH
			if(R_DEPTH(child) >= L_DEPTH(child)) {
#			else
#			ifdef AVL_COUNT
			if(R_COUNT(child) >= L_COUNT(child)) {
#			else
#			error No balancing possible.
#			endif
#			endif
				avlnode->right = child->left;
				if(avlnode->right)
					avlnode->right->parent = avlnode;
				child->left = avlnode;
				avlnode->parent = child;
				*superparent = child;
				child->parent = parent;
#				ifdef AVL_COUNT
				avlnode->count = CALC_COUNT(avlnode);
				child->count = CALC_COUNT(child);
#				endif
#				ifdef AVL_DEPTH
				avlnode->depth = CALC_DEPTH(avlnode);
				child->depth = CALC_DEPTH(child);
#				endif
			} else {
				gchild = child->left;
				avlnode->right = gchild->left;
				if(avlnode->right)
					avlnode->right->parent = avlnode;
				child->left = gchild->right;
				if(child->left)
					child->left->parent = child;
				gchild->left = avlnode;
				if(gchild->left)
					gchild->left->parent = gchild;
				gchild->right = child;
				if(gchild->right)
					gchild->right->parent = gchild;
				*superparent = gchild;
				gchild->parent = parent;
#				ifdef AVL_COUNT
				avlnode->count = CALC_COUNT(avlnode);
				child->count = CALC_COUNT(child);
				gchild->count = CALC_COUNT(gchild);
#				endif
#				ifdef AVL_DEPTH
				avlnode->depth = CALC_DEPTH(avlnode);
				child->depth = CALC_DEPTH(child);
				gchild->depth = CALC_DEPTH(gchild);
#				endif
			}
		break;
		default:
#			ifdef AVL_COUNT
			avlnode->count = CALC_COUNT(avlnode);
#			endif
#			ifdef AVL_DEPTH
			avlnode->depth = CALC_DEPTH(avlnode);
#			endif
		}
		avlnode = parent;
	}
}

#define AVL_CMP_DEFINE_NAMED(n,t) int avl_##n##_cmp(const t a, const t b) { return AVL_CMP(a,b); }
#define AVL_CMP_DEFINE_T(t) AVL_CMP_DEFINE_NAMED(t, t##_t)
#define AVL_CMP_DEFINE(t) AVL_CMP_DEFINE_NAMED(t, t)

AVL_CMP_DEFINE(float)
AVL_CMP_DEFINE(double)

AVL_CMP_DEFINE(char)
AVL_CMP_DEFINE(short)
AVL_CMP_DEFINE(int)
AVL_CMP_DEFINE(long)

AVL_CMP_DEFINE_NAMED(unsigned_char, unsigned char)
AVL_CMP_DEFINE_NAMED(unsigned_short, unsigned short)
AVL_CMP_DEFINE_NAMED(unsigned_int, unsigned int)
AVL_CMP_DEFINE_NAMED(unsigned_long, unsigned long)

AVL_CMP_DEFINE_NAMED(pointer, void *)

#ifdef __GNUC__
__extension__
AVL_CMP_DEFINE_NAMED(long_long, long long)
__extension__
AVL_CMP_DEFINE_NAMED(unsigned_long_long, unsigned long long)
__extension__
AVL_CMP_DEFINE_NAMED(long_double, long double)
#endif

#ifdef HAVE_C99
AVL_CMP_DEFINE_T(int8)
AVL_CMP_DEFINE_T(uint8)
AVL_CMP_DEFINE_T(int16)
AVL_CMP_DEFINE_T(uint16)
AVL_CMP_DEFINE_T(int32)
AVL_CMP_DEFINE_T(uint32)
AVL_CMP_DEFINE_T(int64)
AVL_CMP_DEFINE_T(uint64)

AVL_CMP_DEFINE_T(int_fast8)
AVL_CMP_DEFINE_T(uint_fast8)
AVL_CMP_DEFINE_T(int_fast16)
AVL_CMP_DEFINE_T(uint_fast16)
AVL_CMP_DEFINE_T(int_fast32)
AVL_CMP_DEFINE_T(uint_fast32)
AVL_CMP_DEFINE_T(int_fast64)
AVL_CMP_DEFINE_T(uint_fast64)

AVL_CMP_DEFINE_T(int_least8)
AVL_CMP_DEFINE_T(uint_least8)
AVL_CMP_DEFINE_T(int_least16)
AVL_CMP_DEFINE_T(uint_least16)
AVL_CMP_DEFINE_T(int_least32)
AVL_CMP_DEFINE_T(uint_least32)
AVL_CMP_DEFINE_T(int_least64)
AVL_CMP_DEFINE_T(uint_least64)
#endif

#ifdef HAVE_POSIX
AVL_CMP_DEFINE_T(time)
AVL_CMP_DEFINE_T(size)
AVL_CMP_DEFINE_T(ssize)
AVL_CMP_DEFINE_T(socklen)

int avl_timeval_cmp(const struct timeval *a, const struct timeval *b) {
	int r;
	r = AVL_CMP(a->tv_sec, b->tv_sec);
	if(!r)
		r = AVL_CMP(a->tv_usec, b->tv_usec);
	return r;
}
#endif
